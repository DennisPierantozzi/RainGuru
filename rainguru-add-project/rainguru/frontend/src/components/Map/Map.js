import React from "react";
import Popup from "react-popup";
import L from "leaflet";
import {GeoSearchControl, OpenStreetMapProvider} from "leaflet-geosearch";

import Slider from "../Information/Slider";
import Communication from "../Communication";

export default class Map {

    // static variables that contain everything for the map component
    static map;
    static heatLayers = [];
    static nextLayers = [];
    static currentLayer = new L.LayerGroup();
    static marker;
    static loadedImages = 0;
    static dataURLS = [];
    static minMapBounds = L.latLng(49.18, -4.35);
    static maxMapBounds = L.latLng(55.72, 15.18);
    static minMarkerBounds = L.latLng(50.21, 1.90);
    static maxMarkerBounds = L.latLng(54.71, 9.30);


    /**
     * Preload and cache the images
     */
    static preloadedImages() {
        // get the image urls
        const imageUrls = Communication.getImageUrls();

        // for each url preload the image so it doesn't flicker
        imageUrls.forEach((image) => {
            const newImage = new Image();
            newImage.onload = function () {
                Map.loadedImages += 1;
            }
            newImage.src = image;
            window[image] = newImage;
        });
    }

    /**
     * Generates a list of all generated precipitation images for the coming 100 minutes.
     */
    static setHeatLayers() {
        // heat layer code
        const imageUrls = Communication.getImageUrls();
        const imageBounds = [[54.71, 1.90], [50.21, 9.30]];
        this.heatLayers = [];

        // adds the regular image layers
        for (let i = 0; i < 20; i++) {
            const newUrl = imageUrls[i];
            this.heatLayers.push(L.imageOverlay(newUrl, imageBounds, {opacity: 0.4}));
        }

        // adds the image layers that are used to preload the next layer
        for (let i = 1; i < 20; i++) {
            const newUrl = imageUrls[i];
            this.nextLayers.push(L.imageOverlay(newUrl, imageBounds, {opacity: 0}));
        }
        const zeroUrl = imageUrls[0];
        this.nextLayers.push(L.imageOverlay(zeroUrl, imageBounds, {opacity: 0}));
    }

    /**
     * Shows the requested precipitation image on the map.
     * @param layerIndex the index of the requested precipitation image in the 'heatLayers' list.
     */
    static activateHeatLayer(layerIndex) {
        // remove the layers currently showing
        // then add the current layer
        // and add the next layer with opacity 0 to fix the flickering on safari and firefox
        this.currentLayer.clearLayers();
        this.currentLayer.addLayer(this.heatLayers[layerIndex]);
        this.currentLayer.addLayer(this.nextLayers[layerIndex]);
    }

    /**
     * Renders the map and handles the event of clicking on the map and putting a marker at that location.
     */
    static render() {
        // base layer code
        const baseLayer = L.tileLayer('https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer/tile/{z}/{y}/{x}', {
            attribution: '&copy; <a href="https://www.esri.com/en-us/home">Esri</a>'
        });

        // set up the heat layers
        Map.setHeatLayers();

        // map code
        let map = L.map("map", {
            center: new L.LatLng(52.45, 5.415),
            zoom: 7,
            minZoom: 7,
            maxBounds: L.latLngBounds(Map.minMapBounds, Map.maxMapBounds),
            maxBoundsViscosity: 0.9,
        });

        // add layers to the map and the scale
        baseLayer.addTo(map);
        this.currentLayer.addTo(map);
        L.control.scale({imperial: false}).addTo(map);

        // load the first rain image and center the map to the Netherlands
        this.activateHeatLayer(0);

        // store the map
        Map.map = map;

        // add an error message for the users current location
        map.on('locationerror', function (e) {
            Popup.alert("User denied access to their location", "Permission denied");
        });

        // add location finding handling to the map
        map.on('locationfound', function (e) {
            Map.markerHandling(e, false, true);
        });

        // add clicking to place a marker to the map
        map.on('click', function (e) {
            Map.markerHandling(e, false, false);
        });

        // add the search functionality to the map
        Map.addMapSearch();

        // When a location has been found this event will be triggered
        map.on('geosearch/showlocation', function (e) {
            Map.markerHandling(e, true, false);
        });
    }

    /**
     * Add the search functionality to the map.
     */
    static addMapSearch() {
        // Adds searching functionality to the map
        const search = new GeoSearchControl({
            provider: new OpenStreetMapProvider(),
            position: 'topleft',
            animateZoom: false,
            retainZoomLevel: true,
            marker: {
                icon: new L.Icon.Default(),
                draggable: false,
            },
            searchLabel: 'Search',
            updateMap: true,
        });
        Map.map.addControl(search);

        // Move the search bar to the top bar
        document.getElementById('locationField').appendChild(
            document.querySelector(".geosearch")
        );
    }

    // method for the map.on events
    /**
     * Handles the placement of the map markers in various situations.
     *
     * @param e the event.
     * @param search a boolean representing if the method was called for a geosearch event.
     * @param found a boolean representing if the method was called for when a location was found (otherwise it is on click).
     */
    static markerHandling(e, search, found) {
        // the location of where we want to try to place a marker
        let latitude;
        let longitude;
        if (search) {
            latitude = e.location.y;
            longitude = e.location.x;
        } else {
            latitude = e.latlng.lat;
            longitude = e.latlng.lng;
        }

        // check if the location is inside the bounds
        if (latitude > Map.maxMarkerBounds.lat || latitude < Map.minMarkerBounds.lat || longitude < Map.minMarkerBounds.lng || longitude > Map.maxMarkerBounds.lng) {
            Popup.alert("Your selected location is outside the bounds of the precipitation map", "Location out of bounds");
            if (search) {
                Map.map.setView(Map.marker !== undefined ? Map.marker.getLatLng() : L.latLng(52.45, 5.415), 8);
            }
            return;
        }

        // remove any existing markers
        if (Map.marker !== undefined) {
            Map.map.removeLayer(Map.marker);
        }

        // add the marker
        Map.marker = L.marker([latitude, longitude]).addTo(Map.map);
        // go to the new marker, this is handled differently depending on what event called the method
        if (search) {
            Map.map.setZoom(8);
        } else {
            Map.map.flyTo(L.latLng(latitude, longitude), found ? Math.max(8, Map.map.getZoom()) : Map.map.getZoom());
        }
        Slider.showPrecipitationData(latitude, longitude);
    }
}
