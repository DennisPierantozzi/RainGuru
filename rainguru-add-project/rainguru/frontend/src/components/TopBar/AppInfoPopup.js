import React from 'react';

/**
 * Generates a div for a popup where the content will be rendered in later.
 * @param props contains state of a particular popup, method to change the state, and text to display
 * @returns a div for a sidebar popup.
 */
function AppInfoPopup(props) {

    return (props.state) ? (
        <div className="popup" data-testid="AppInfoPopupContainer" onClick={(e) => props.changePopup(e, false)}>
            <div className="popup-box" data-testid="AppInfoPopupBox" onClick={(e) => props.changePopup(e, true)}>
                <div className="popup-content">
                    <div className="popup-information">
                        {props.children}
                    </div>
                    <div className="popup-close">
                        <button className="close-button" data-testid="AppInfoCloseButton" onClick={(e) => props.changePopup(e, false)}>
                            <img className="ratioImage" src="../../../static/images/3x1.png"/>
                            <p>Close</p>
                        </button>
                    </div>
                </div>
            </div>
        </div>
    ) : "";
}

export default AppInfoPopup