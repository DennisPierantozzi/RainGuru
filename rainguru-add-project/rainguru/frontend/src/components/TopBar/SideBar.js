import React from 'react';
import PastDataSelector from "./PastDataSelector";

export default function SideBar({pastDataSelectorUpdateProp, loadingData, displayData, compareData, displayComparedData, setShowCompare, setShowSidebar}) {

    return (
        <div className="Sidebar-shown">
            <div className='container-sidebar-menu'>
                <div className="Menu-title">
                    <div>RainGuru <br/> <span className="Menu-subtitle">Weather forecast app</span></div>
                    <img src="../../static/images/TUDelft_logo_rgb.webp" width="10vw" height="13vh" alt="TUDelft Logo" />
                </div>
                <PastDataSelector updateProp={pastDataSelectorUpdateProp} loadingData={loadingData} displayData={displayData} compareData={compareData} 
                setShowSidebar={setShowSidebar}/>

                <div className='Bottom-privacy-statement'>
                    <p>This website does not store privacy sensitive information, except for the location of the user during usage if permitted.</p>
                </div>
            </div> 
        </div>
    )
}