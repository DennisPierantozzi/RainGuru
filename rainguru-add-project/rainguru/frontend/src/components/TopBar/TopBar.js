import React, {useState} from 'react';
import {FaBars} from 'react-icons/fa';
import {HiOutlineDatabase} from 'react-icons/hi';
import {ImLocation} from 'react-icons/im';
import {SideBarData} from "./SideBarData";
import AppInfoPopup from "./AppInfoPopup";
import Map from "../Map/Map";
import PastDataSelector from "./PastDataSelector";

export default function TopBar({pastDataSelectorUpdateProp, loadingData, displayData}) {

    /**
     * initializing a state for sidebar and function to update it. Initially, state is set to false.
     * This state is to keep track of whether the sidebar is open or not
     */
    const [sidebar, setSidebar] = useState(false)

    /**
     * toggles the value to the opposite state.
     */
    const showSidebar = () => setSidebar(!sidebar)

    /**
     * displays precipitation information over user's current location and also leave a marker at that location.
     */
    const showCurrentLocation = () => {
        Map.map.locate();
    }

    /**
     * displays precipitation information at TU Delft and also marks the location.
     */
    const showTUDelftLocation = () => {
        const latlngPoint = new L.LatLng(52.0022,4.3736);
        Map.map.fireEvent('click',{latlng:latlngPoint});
    }

    /**
     * closes the pastData menu.
     */
    const closeMenus = () => {
        setPastDataSelector(false);
    }

    /**
     * initializing a state for pastDataSelector and function to update it. Initially, state is set to false.
     * This state is to keep track of whether pastData menu is open or not.
     */
    const [pastDataSelector, setPastDataSelector] = useState(false);

    /**
     * sets pastDataSelector state depending on the current state and closes menu if the current state is false.
     */
    const showPastDataSelector = () => {
        if (!pastDataSelector) {
            closeMenus();
            setPastDataSelector(true);
        } else {
            setPastDataSelector(false);
        }
    }

    /**
     * Initializing five states and functions to update them. Initial states are set to false for all of them.
     * These states are to keep track of whether or not a popup is displayed.
     */
    const [appInfoPopup, setAppInfoPopup] = useState(false)
    const [contributorPopup, setContributorPopup] = useState(false)
    const [disclaimerPopup, setDisclaimerPopup] = useState(false)
    const [privacyPopup, setPrivacyPopup] = useState(false)
    const [acknowledgementPopup, setAcknowledgementPopup] = useState(false);

    /**
     * Updates application information popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changeAppInfoPopup = (e, bool) => {
        e.stopPropagation();
        setAppInfoPopup(bool);
    }

    /**
     * Updates disclaimer and liability popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changeDisclaimerPopup = (e, bool) => {
        e.stopPropagation();
        setDisclaimerPopup(bool);
    }

    /**
     * Updates privacy statement popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changePrivacyPopup = (e, bool) => {
        e.stopPropagation();
        setPrivacyPopup(bool);
    }

    /**
     * Updates contributor popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changeContributorPopup = (e, bool) => {
        e.stopPropagation();
        setContributorPopup(bool);
    }

    /**
     * Updates acknowledgement popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changeAcknowledgementPopup = (e, bool) => {
        e.stopPropagation();
        setAcknowledgementPopup(bool);
    }

    /**
     * Renders the content of the TopBar div at the top of the screen. This also contains sidebar.
     * @returns the rendered content of the topbar div.
     */
    return (
        <div className="topBarContents" data-testid="topBar-container">
            <div className="Sidebar" data-testid="sidebar-menu">
                <img className="ratioImage" src="../../../static/images/1x1.png"/>
                <FaBars className="Sidebar-icon" onClick={showSidebar} size="60%"/>
            </div>
            <div className="Location">
                <img className="ratioImage" src="../../../static/images/1x1.png"/>
                <ImLocation className="Location-icon" data-testid="user-location" onClick={showCurrentLocation} size="60%"/>
            </div>
            <div className="TUDelft-logo">
                <img className="ratioImage" src="../../../static/images/1x1.png"/>
                <img src="../../../static/images/TUDelftButton.png" alt="TU Delft logo"
                     className="TUDelft-logo-icon" data-testid="tudelft-location" onClick={showTUDelftLocation}></img>
            </div>
            <div className="Searchbar" id="locationField"></div>

            <nav className={sidebar ? "Sidebar-shown" : "hideElement"}>
                <ul className="Sidebar-items">
                    <li className={pastDataSelector ? "active-menu-item" : "SideBar-item"} onClick={showPastDataSelector}>
                        <span className="item-icon">
                            <img className="ratioImage" src="../../../static/images/1x1.png"/>
                            <HiOutlineDatabase className="sidebar-element-icon" size="50%"/>
                        </span>
                        <span className="item-title">Display previous data</span>
                    </li>
                    <li className={pastDataSelector ? "open-menu" : "hideElement"}>
                        <PastDataSelector updateProp={pastDataSelectorUpdateProp} loadingData={loadingData} displayData={displayData}/>
                    </li>
                    <li className={pastDataSelector ? "hideElement" : "sidebar-splitter"} id="sidebar-splitter"></li>
                    {SideBarData.map((item, index) => {
                        return (
                            <li key={index} className={item.className} data-testid={item.id}
                                onClick={item.title === "Application information" ? (e) => changeAppInfoPopup(e, true) :
                                    item.title === "Contributors" ? (e) => changeContributorPopup(e, true) :
                                    item.title === "Disclaimer & Liability" ? (e) => changeDisclaimerPopup(e, true) :
                                    item.title === "Acknowledgement" ? (e) => changeAcknowledgementPopup(e, true) :
                                        (e) => changePrivacyPopup(e, true)}>
                                <span className="item-icon">
                                    <img className="ratioImage" src="../../../static/images/1x1.png"/>
                                    {item.icon}
                                </span>
                                <span className="item-title">{item.title}</span>
                                <div className="popup-container" data-testid={item.id + "-popup-container"}>
                                    <AppInfoPopup
                                        state={item.title === "Application information" ? appInfoPopup : item.title === "Disclaimer & Liability" ? disclaimerPopup:
                                            item.title === "Contributors" ? contributorPopup : item.title === "Acknowledgement" ? acknowledgementPopup : privacyPopup}
                                        changePopup={item.title === "Application information" ? changeAppInfoPopup :
                                            item.title === "Contributors" ? changeContributorPopup:
                                            item.title === "Disclaimer & Liability" ?  changeDisclaimerPopup:
                                            item.title === "Acknowledgement" ? changeAcknowledgementPopup: changePrivacyPopup}>
                                        <div id="popup-text"data-testid={item.id + "-popuptext-topbar"}>{item.popupText}</div>
                                    </AppInfoPopup>
                                </div>
                            </li>
                        )
                    })}
                </ul>
            </nav>
        </div>
    )
}
