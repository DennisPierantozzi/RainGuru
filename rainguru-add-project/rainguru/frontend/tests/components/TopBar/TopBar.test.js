import React from "react";
import {fireEvent, render, screen} from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import TopBar  from "../../../src/components/TopBar/TopBar";
import Map from "../../../src/components/Map/Map";

jest.mock("../../../src/components/TopBar/PastDataSelector");

// disable certain warnings
const warnings = console.error.bind(console.error)
beforeAll(() => { console.error = errormessage => {
    !errormessage.toString().includes("Warning: `%s` uses `getDerivedStateFromProps` but its initial state is %s. " +
        "This is not recommended. Instead, define the initial state by assigning an object to `this.state` in the " +
        "constructor of `%s`. This ensures that `getDerivedStateFromProps` arguments have a consistent shape.%s")
        && !errormessage.toString().includes("Warning: %s.getDerivedStateFromProps(): " +
        "A valid state object (or null) must be returned. You have returned undefined.%s")
        && warnings(errormessage);
}})
afterAll(() => { console.error = warnings })

afterEach(() => {
    jest.clearAllMocks();
});

describe("TopBar basic tests", () => {
    it("matches snapshot", () => {
        const state = renderer.create(<TopBar/>).toJSON();
        expect(state).toMatchSnapshot();
    });

    it("renders Topbar properly", () => {
        render(<TopBar/>);
        const element = screen.getByTestId("topBar-container");
        expect(element).toBeTruthy();
    });
});

describe("icons on top bar tests", () => {
    it("user location search works as intended", () => {
        render(<TopBar loadingData={true} updateProp={0} displayData={jest.fn()}/>)
        const userLocationIcon = screen.getByTestId("user-location");
        Map.map = {locate: jest.fn()};
        expect(Map.map.locate).toBeCalledTimes(0);
        // When the location icon is clicked, method to show user's current location should be called
        fireEvent.click(userLocationIcon);
        expect(Map.map.locate).toBeCalledTimes(1);
    });

    it("tu delft icon button works as intended", () => {
        render(<TopBar loadingData={true} updateProp={0} displayData={jest.fn()}/>)
        const userLocationIcon = screen.getByTestId("tudelft-location");
        Map.map = {fireEvent: jest.fn()};
        expect(Map.map.fireEvent).toBeCalledTimes(0);
        // When the location icon is clicked, method to show user's current location should be called
        fireEvent.click(userLocationIcon);
        expect(Map.map.fireEvent).toBeCalledTimes(1);
    });
});

describe("items in the sidebar works as intended", () => {

    it("Application information test", () => {
        render(<TopBar loadingData={true} updateProp={0} displayData={jest.fn()}/>)
        const AppInfoPopupContainer = screen.getByTestId("application-info-popup-container");
        expect(AppInfoPopupContainer).toBeEmptyDOMElement();

        fireEvent.click(AppInfoPopupContainer);
        // when the app info sidebar menu is clicked, AppInfoPopupContainer should no longer be an empty dom element.
        expect(AppInfoPopupContainer.childNodes.length).toBe(1);
        // text of a div inside of AppInfoPopupContainer that appears when clicking on AppInfoPopupContainer
        // should also equal to the corresponding app info popup text set in SideBarData.js
        const title = screen.getByTestId("application-info-popuptext-topbar");
        expect(title.firstChild).toStrictEqual(screen.getByTestId("application-info-popuptext-sidebardata"));
    });

    it("Contributors test", () => {
        render(<TopBar loadingData={true} updateProp={0} displayData={jest.fn()}/>)
        const ContributorPopupContainer = screen.getByTestId("contributor-popup-container");
        expect(ContributorPopupContainer).toBeEmptyDOMElement();

        fireEvent.click(ContributorPopupContainer);
        // when the contributor sidebar menu is clicked, ContributorPopupContainer should no longer be an empty dom element.
        expect(ContributorPopupContainer.childNodes.length).toBe(1);
        // text of a div inside of ContributorPopupContainer that appears when clicking on ContributorPopupContainer
        // should also equal to the corresponding popup text set in SideBarData.js
        const title = screen.getByTestId("contributor-popuptext-topbar");
        expect(title.firstChild).toStrictEqual(screen.getByTestId("contributor-popuptext-sidebardata"));
    });

    it("Disclaimer test", () => {
        render(<TopBar loadingData={true} updateProp={0} displayData={jest.fn()}/>)
        const DisclaimerPopupContainer = screen.getByTestId("disclaimer-popup-container");
        expect(DisclaimerPopupContainer).toBeEmptyDOMElement();

        fireEvent.click(DisclaimerPopupContainer);
        // when the disclaimer sidebar menu is clicked, DisclaimerPopupContainer should no longer be an empty dom element.
        expect(DisclaimerPopupContainer.childNodes.length).toBe(1);
        // text of a div inside of DisclaimerPopupContainer that appears when clicked on DisclaimerPopupContainer
        // should also equal to the corresponding disclaimer popup text set in SideBarData.js
        const title = screen.getByTestId("disclaimer-popuptext-topbar");
        expect(title.firstChild).toStrictEqual(screen.getByTestId("disclaimer-popuptext-sidebardata"));
    });

    it("privacy info test", () => {
        render(<TopBar loadingData={true} updateProp={0} displayData={jest.fn()}/>)
        const PrivacyPopupContainer = screen.getByTestId("privacy-popup-container");
        expect(PrivacyPopupContainer).toBeEmptyDOMElement();

        fireEvent.click(PrivacyPopupContainer);
        // when the privacy info sidebar menu is clicked, PrivacyPopupContainer should no longer be an empty dom element.
        expect(PrivacyPopupContainer.childNodes.length).toBe(1);
        // text of a div inside of PrivacyPopupContainer that appears when PrivacyPopupContainer gets clicked
        // should also equal to the corresponding popup text set in SideBarData.js
        const title = screen.getByTestId("privacy-popuptext-topbar");
        expect(title.firstChild).toStrictEqual(screen.getByTestId("privacy-popuptext-sidebardata"));
    });

    it("acknowledgement test", () => {
        render(<TopBar loadingData={true} updateProp={0} displayData={jest.fn()}/>)
        const AcknowledgementPopupContainer = screen.getByTestId("acknowledgement-popup-container");
        expect(AcknowledgementPopupContainer).toBeEmptyDOMElement();

        fireEvent.click(AcknowledgementPopupContainer);
        // when the acknowledgement menu is clicked, AcknowledgementPopupContainer should no longer be an empty dom element.
        expect(AcknowledgementPopupContainer.childNodes.length).toBe(1);
        // text of a div inside of AcknowledgementPopupContainer that appears when AcknowledgementPopupContainer is
        // clicked should also equal to the corresponding acknowledgement popup text defined in SideBarData.js
        const title = screen.getByTestId("acknowledgement-popuptext-topbar");
        expect(title.firstChild).toStrictEqual(screen.getByTestId("acknowledgement-popuptext-sidebardata"));
    });
});