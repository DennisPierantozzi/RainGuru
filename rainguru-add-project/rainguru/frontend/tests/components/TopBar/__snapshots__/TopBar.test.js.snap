// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar basic tests matches snapshot 1`] = `
<div
  className="topBarContents"
  data-testid="topBar-container"
>
  <div
    className="Sidebar"
    data-testid="sidebar-menu"
  >
    <img
      className="ratioImage"
      src="../../../static/images/1x1.png"
    />
    <svg
      className="Sidebar-icon"
      fill="currentColor"
      height="60%"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="60%"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      />
    </svg>
  </div>
  <div
    className="Location"
  >
    <img
      className="ratioImage"
      src="../../../static/images/1x1.png"
    />
    <svg
      className="Location-icon"
      data-testid="user-location"
      fill="currentColor"
      height="60%"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      version="1.1"
      viewBox="0 0 16 16"
      width="60%"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 0c-2.761 0-5 2.239-5 5 0 5 5 11 5 11s5-6 5-11c0-2.761-2.239-5-5-5zM8 8c-1.657 0-3-1.343-3-3s1.343-3 3-3 3 1.343 3 3-1.343 3-3 3z"
      />
    </svg>
  </div>
  <div
    className="TUDelft-logo"
  >
    <img
      className="ratioImage"
      src="../../../static/images/1x1.png"
    />
    <img
      alt="TU Delft logo"
      className="TUDelft-logo-icon"
      data-testid="tudelft-location"
      onClick={[Function]}
      src="../../../static/images/TUDelftButton.png"
    />
  </div>
  <div
    className="Searchbar"
    id="locationField"
  />
  <nav
    className="hideElement"
  >
    <ul
      className="Sidebar-items"
    >
      <li
        className="SideBar-item"
        onClick={[Function]}
      >
        <span
          className="item-icon"
        >
          <img
            className="ratioImage"
            src="../../../static/images/1x1.png"
          />
          <svg
            className="sidebar-element-icon"
            fill="none"
            height="50%"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="50%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 7v10c0 2.21 3.582 4 8 4s8-1.79 8-4V7M4 7c0 2.21 3.582 4 8 4s8-1.79 8-4M4 7c0-2.21 3.582-4 8-4s8 1.79 8 4m0 5c0 2.21-3.582 4-8 4s-8-1.79-8-4"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            />
          </svg>
        </span>
        <span
          className="item-title"
        >
          Display previous data
        </span>
      </li>
      <li
        className="hideElement"
      />
      <li
        className="sidebar-splitter"
        id="sidebar-splitter"
      />
      <li
        className="SideBar-item"
        data-testid="application-info"
        onClick={[Function]}
      >
        <span
          className="item-icon"
        >
          <img
            className="ratioImage"
            src="../../../static/images/1x1.png"
          />
          <svg
            className="sidebar-element-icon"
            fill="currentColor"
            height="50%"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 20 20"
            width="50%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7-4a1 1 0 11-2 0 1 1 0 012 0zM9 9a1 1 0 000 2v3a1 1 0 001 1h1a1 1 0 100-2v-3a1 1 0 00-1-1H9z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="item-title"
        >
          Application information
        </span>
        <div
          className="popup-container"
          data-testid="application-info-popup-container"
        />
      </li>
      <li
        className="SideBar-item"
        data-testid="contributor"
        onClick={[Function]}
      >
        <span
          className="item-icon"
        >
          <img
            className="ratioImage"
            src="../../../static/images/1x1.png"
          />
          <svg
            className="sidebar-element-icon"
            fill="currentColor"
            height="50%"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="50%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M337.454 232c33.599 0 61.092-27.002 61.092-60 0-32.997-27.493-60-61.092-60s-61.09 27.003-61.09 60c0 32.998 27.491 60 61.09 60zm-162.908 0c33.599 0 61.09-27.002 61.09-60 0-32.997-27.491-60-61.09-60s-61.092 27.003-61.092 60c0 32.998 27.493 60 61.092 60zm0 44C126.688 276 32 298.998 32 346v54h288v-54c0-47.002-97.599-70-145.454-70zm162.908 11.003c-6.105 0-10.325 0-17.454.997 23.426 17.002 32 28 32 58v54h128v-54c0-47.002-94.688-58.997-142.546-58.997z"
            />
          </svg>
        </span>
        <span
          className="item-title"
        >
          Contributors
        </span>
        <div
          className="popup-container"
          data-testid="contributor-popup-container"
        />
      </li>
      <li
        className="SideBar-item"
        data-testid="disclaimer"
        onClick={[Function]}
      >
        <span
          className="item-icon"
        >
          <img
            className="ratioImage"
            src="../../../static/images/1x1.png"
          />
          <svg
            className="sidebar-element-icon"
            fill="none"
            height="50%"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="50%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 6C12.5523 6 13 6.44772 13 7V13C13 13.5523 12.5523 14 12 14C11.4477 14 11 13.5523 11 13V7C11 6.44772 11.4477 6 12 6Z"
              fill="currentColor"
            />
            <path
              d="M12 16C11.4477 16 11 16.4477 11 17C11 17.5523 11.4477 18 12 18C12.5523 18 13 17.5523 13 17C13 16.4477 12.5523 16 12 16Z"
              fill="currentColor"
            />
            <path
              clipRule="evenodd"
              d="M12 2C6.47715 2 2 6.47715 2 12C2 17.5228 6.47715 22 12 22C17.5228 22 22 17.5228 22 12C22 6.47715 17.5228 2 12 2ZM4 12C4 16.4183 7.58172 20 12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="item-title"
        >
          Disclaimer & Liability
        </span>
        <div
          className="popup-container"
          data-testid="disclaimer-popup-container"
        />
      </li>
      <li
        className="SideBar-item"
        data-testid="privacy"
        onClick={[Function]}
      >
        <span
          className="item-icon"
        >
          <img
            className="ratioImage"
            src="../../../static/images/1x1.png"
          />
          <svg
            className="sidebar-element-icon"
            fill="currentColor"
            height="50%"
            role="img"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="50%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title />
            <path
              d="M2.968 11.583h1.274v-3.82A7.76 7.76 0 0 1 12.005 0a7.76 7.76 0 0 1 7.762 7.763v3.783c-.018.01-.037.028-.056.037l-.01.01-.008.009h-.01l-.01.01-.009.009H19.636l-.018.018h-.02l-.018.01h-.01l-.009.01-.009.009h-.01l-.009.009-.009.01-.01.009-.009.009-.028.019-.019.01-.028.018-.018.01-.02.009-.027.018-.019.01-.01.009-.027.019-.02.01-.046.027-.019.01-.018.009-.02.01h-.008l-.057.027h-.019c-.018.01-.037.02-.065.038h-.01l-.009.01-.028.018-.018.01-.029.018-.018.01h-.01l-.028.018-.018.01-.02.009c-.018.01-.046.019-.065.028l-.018.01-.02.009-.037.018-.037.02-.047.018-.047.019-.019.009-.037.019-.019.01c-1.545.739-4.017 1.516-8.708 1.853-3.362.244-5.403 1.723-6.724 3.502zm4.842 0h8.371v-3.82a4.184 4.184 0 0 0-4.186-4.186A4.184 4.184 0 0 0 7.81 7.763zm13.222 1.461V24H5.572c1.704-.946 2.968-.852 5.075-.787 2.865.094 6.03-1.105 7.585-2.696 1.554-1.592-.14-.375-1.901.074-1.76.45-5.17.497-7.454-.103 7.173.094 9.973-2.219 11.555-4.307 1.583-2.079-.683-.365-2.153.356-1.47.72-4.036 1.227-6.864.852 4.27-.01 7.52-2.144 9.607-4.345z"
            />
          </svg>
        </span>
        <span
          className="item-title"
        >
          Privacy statement
        </span>
        <div
          className="popup-container"
          data-testid="privacy-popup-container"
        />
      </li>
      <li
        className="SideBar-item"
        data-testid="acknowledgement"
        onClick={[Function]}
      >
        <span
          className="item-icon"
        >
          <img
            className="ratioImage"
            src="../../../static/images/1x1.png"
          />
          <svg
            className="sidebar-element-icon"
            fill="currentColor"
            height="50%"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="50%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M448 360V24c0-13.3-10.7-24-24-24H96C43 0 0 43 0 96v320c0 53 43 96 96 96h328c13.3 0 24-10.7 24-24v-16c0-7.5-3.5-14.3-8.9-18.7-4.2-15.4-4.2-59.3 0-74.7 5.4-4.3 8.9-11.1 8.9-18.6zM128 134c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm0 64c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm253.4 250H96c-17.7 0-32-14.3-32-32 0-17.6 14.4-32 32-32h285.4c-1.9 17.1-1.9 46.9 0 64z"
            />
          </svg>
        </span>
        <span
          className="item-title"
        >
          Acknowledgement
        </span>
        <div
          className="popup-container"
          data-testid="acknowledgement-popup-container"
        />
      </li>
    </ul>
  </nav>
</div>
`;
